File: backend/helpers/SessionManager/session_manager.go
package SessionManager

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"time"

	"eunity.com/backend-main/helpers/DBManager"
	"eunity.com/backend-main/models"
	"github.com/gin-gonic/gin"
	"github.com/go-redis/redis/v8"
	"github.com/google/uuid"
	"github.com/joho/godotenv"
	"go.mongodb.org/mongo-driver/bson"
)

// retrieves session data using the session_id stored in the cookies
var ValKey bool
var rdb *redis.Client
var ctx = context.Background()

// https only must be true in production
var HTTPS_only bool

// must be set to eunityusa.com in production
var Cookie_Host string

func init() {
	godotenv.Load()
	godotenv.Load()
	HTTPS_only = os.Getenv("HTTPS_ONLY") == "true"
	Cookie_Host = os.Getenv("COOKIE_ACCEPT_HOST")
	valkey := os.Getenv("VALKEY_ENABLED")
	if valkey == "true" {
		ValKey = true
		rdb = redis.NewClient(&redis.Options{
			Addr:     "localhost:6379",
			Password: "123456", // no password set
			DB:       0,        // use default DB
		})

		_, err := rdb.Ping(ctx).Result()
		if err != nil {
			fmt.Println("Error connecting to Valkey")
			ValKey = false
		}
	} else {
		ValKey = false
	}
}

func Get_Session(session_id string) (models.Session, error) {
	switch ValKey {
	case true:
		//get session id from valkey
		session, err := rdb.Get(ctx, session_id).Result()
		if err != nil {
			return models.Session{}, err
		}
		var sessionData models.Session
		err = json.Unmarshal([]byte(session), &sessionData)
		if err != nil {
			return models.Session{}, err
		}
		return sessionData, nil
	default:
		//get session id from mongo db
		session := DBManager.DB.Collection("session_ids").FindOne(context.Background(), bson.M{"session_id": session_id})
		if session.Err() != nil {
			return models.Session{}, session.Err()
		}

		//get session data
		var sessionData models.Session
		err := session.Decode(&sessionData)
		if err != nil {
			return models.Session{}, err
		}
		return sessionData, nil
	}
}

func Create_Session(user_id string, c *gin.Context) (models.Session, error) {
	cookie_max_age := 365 * 24 * 60 * 60 //1 year
	session := models.Session{
		Session_id: uuid.New().String(),
		Created_at: time.Now().Unix(),
		//we don't deny access if expired we just rotate session_ids
		Expires_at:  time.Now().Unix() + 3600*6, //6 hours
		User_id:     user_id,
		Permissions: []string{"user"}, //default permissions
	}
	jsonSession, err := json.Marshal(session)
	if err != nil {
		return session, err
	}
	switch ValKey {
	case true:
		fmt.Println("Creating session in Valkey")
		//add session to Valkey as map from session_id to session in json
		err = rdb.Set(ctx, session.Session_id, jsonSession, 0).Err()
		if err != nil {
			return session, err
		}

		c.SetCookie("session_id", session.Session_id, cookie_max_age, "/", Cookie_Host, HTTPS_only, true)
		return session, nil

	default:
		//create session id in mongo db
		_, err = DBManager.DB.Collection("session_ids").InsertOne(context.Background(), session)
		if err != nil {
			return session, err
		}
		c.SetCookie("session_id", session.Session_id, cookie_max_age, "/", Cookie_Host, HTTPS_only, true)
		return session, nil
	}

}

func Create_Developer_Session(user_id string, c *gin.Context) (models.Session, error) {
	cookie_max_age := 365 * 24 * 60 * 60 //1 year
	session := models.Session{
		Session_id: uuid.New().String(),
		Created_at: time.Now().Unix(),
		//we don't deny access if expired we just rotate session_ids
		Expires_at:  time.Now().Unix() + 3600*6, //6 hours
		User_id:     user_id,
		Permissions: []string{"developer"}, //default permissions
	}
	jsonSession, err := json.Marshal(session)
	if err != nil {
		return session, err
	}
	switch ValKey {
	case true:
		fmt.Println("Creating session in Valkey")
		//add session to Valkey as map from session_id to session in json
		err = rdb.Set(ctx, session.Session_id, jsonSession, 0).Err()
		if err != nil {
			return session, err
		}

		c.SetCookie("session_id", session.Session_id, cookie_max_age, "/", Cookie_Host, HTTPS_only, true)
		return session, nil

	default:
		//create session id in mongo db
		_, err = DBManager.DB.Collection("session_ids").InsertOne(context.Background(), session)
		if err != nil {
			return session, err
		}
		c.SetCookie("session_id", session.Session_id, cookie_max_age, "/", Cookie_Host, HTTPS_only, true)
		return session, nil
	}

}

func Delete_Session(session_id string) error {
	switch ValKey {
	case true:
		//delete session from valkey
		_, err := rdb.Del(ctx, session_id).Result()
		if err != nil {
			return err
		}
		return nil
	default:
		//delete session from mongo db
		_, err := DBManager.DB.Collection("session_ids").DeleteOne(context.Background(), bson.M{session_id: bson.M{"$exists": true}})
		if err != nil {
			return err
		}
		return nil
	}
}

func AuthRequired() gin.HandlerFunc {
	return func(c *gin.Context) {

		//must change this to domain name in production
		c.Writer.Header().Set("Access-Control-Allow-Origin", "*")

		//check cookies
		session_id, err := c.Cookie("session_id")
		if err != nil {
			c.JSON(401, gin.H{
				"response": "Unauthorized",
			})
			c.Abort()
			return
		}

		switch ValKey {
		case true:

			//check if session_id exists in valkey
			session, err := Get_Session(session_id)
			if err != nil {
				c.JSON(401, gin.H{
					"response": "Unauthorized",
				})
				c.Abort()
				return
			}

			expires_at := time.Unix(session.Expires_at, 0)

			if time.Now().After(expires_at) {
				old_session_id := session.Session_id
				session, err = Create_Session(session.User_id, c)
				if err != nil {
					c.JSON(500, gin.H{
						"response": "Error creating session",
					})
					c.Abort()
					return
				}

				//delete old session
				err = Delete_Session(old_session_id)
				if err != nil {
					c.JSON(500, gin.H{
						"response": "Error rotating old session",
					})
					c.Abort()
					return
				}

			}

			//set all session data in c keys for easy access
			c.Set("session_id", session.Session_id)
			c.Set("created_at", session.Created_at)
			c.Set("expires_at", session.Expires_at)
			c.Set("user_id", session.User_id)
			c.Set("permissions", session.Permissions)

			return

		default:
			//check if session_id exists in session_ids collection
			session, err := Get_Session(session_id)
			if err != nil {
				c.JSON(401, gin.H{
					"response": "Unauthorized",
				})
				c.Abort()
				return
			}

			//check if session is expired
			expires_at := time.Unix(session.Expires_at, 0)

			if time.Now().After(expires_at) {
				old_session_id := session.Session_id
				session, err = Create_Session(session.User_id, c)
				if err != nil {
					c.JSON(500, gin.H{
						"response": "Error creating session",
					})
					c.Abort()
					return
				}

				//delete old session
				err = Delete_Session(old_session_id)
				if err != nil {
					c.JSON(500, gin.H{
						"response": "Error rotating old session",
					})
					c.Abort()
					return
				}
			}

			//set all session data in c keys for easy access
			c.Set("session_id", session.Session_id)
			c.Set("created_at", session.Created_at)
			c.Set("expires_at", session.Expires_at)
			c.Set("user_id", session.User_id)
			c.Set("permissions", session.Permissions)
			c.Next()
		}
		c.Next()
	}
}

File: python/helpers/session_manager.py
from fastapi import FastAPI, HTTPException, Cookie
from fastapi.responses import JSONResponse
from fastapi.requests import Request
from fastapi.security.utils import get_authorization_scheme_param
from pydantic import BaseModel
from typing import Optional
import uuid
import json
import os
import time
from db_manager import DBManager
from redis import Redis, ConnectionPool
from pymongo import MongoClient
from models.session import Session
from dotenv import load_dotenv


load_dotenv()



# Environment variables
HTTPS_ONLY = os.environ.get("HTTPS_ONLY", "false").lower() == "true"
COOKIE_HOST = os.environ.get("COOKIE_ACCEPT_HOST", "")
VALKEY_ENABLED = os.environ.get("VALKEY_ENABLED", "false").lower() == "true"
VALKEY_URL = os.environ.get("VALKEY_URL", "")
VALKEY_PASS = os.environ.get("VALKEY_PASS", "")

# Redis client
if VALKEY_ENABLED:
    redis_pool = ConnectionPool(host=VALKEY_URL, password=VALKEY_PASS, db=0)
    rdb = Redis(connection_pool=redis_pool)
else:
    rdb = None

# MongoDB client
db = DBManager.client




async def get_session(session_id: str) -> Optional[Session]:
    if VALKEY_ENABLED:
        session = await rdb.get(session_id)
        if session:
            return Session(**json.loads(session))
    else:
        collection = db["session_ids"]
        session = collection.find_one({"session_id": session_id})
        if session:
            return Session(**session)
    return None

async def create_session(user_id: str, request: Request) -> Session:
    session_id = str(uuid.uuid4())
    session = Session(
        session_id=session_id,
        created_at=int(time.time()),
        expires_at=int(time.time() + 3600 * 6),
        user_id=user_id,
        permissions=["user"],
    )
    if VALKEY_ENABLED:
        await rdb.set(session_id, json.dumps(session.dict()))
    else:
        collection = db["session_ids"]
        collection.insert_one(session.dict())
    response = JSONResponse(status_code=200, content={"response": "Session created"})
    response.set_cookie("session_id", session_id, max_age=31536000, secure=HTTPS_ONLY, httponly=True)
    return session

async def create_developer_session(user_id: str, request: Request) -> Session:
    session_id = str(uuid.uuid4())
    session = Session(
        session_id=session_id,
        created_at=int(time.time()),
        expires_at=int(time.time() + 3600 * 6),
        user_id=user_id,
        permissions=["developer"],
    )
    if VALKEY_ENABLED:
        await rdb.set(session_id, json.dumps(session.dict()))
    else:
        collection = db["session_ids"]
        collection.insert_one(session.dict())
    response = JSONResponse(status_code=200, content={"response": "Session created"})
    response.set_cookie("session_id", session_id, max_age=31536000, secure=HTTPS_ONLY, httponly=True)
    return session

async def delete_session(session_id: str) -> None:
    if VALKEY_ENABLED:
        await rdb.delete(session_id)
    else:
        collection = db["session_ids"]
        collection.delete_one({"session_id": session_id})


async def auth_required(request: Request, call_next):
    session_id = request.cookies.get("session_id")
    if not session_id:
        return JSONResponse(status_code=401, content={"response": "Unauthorized"})

    session = await get_session(session_id)
    if not session:
        return JSONResponse(status_code=401, content={"response": "Unauthorized"})

    expires_at = session.expires_at
    if time.time() > expires_at:
        old_session_id = session.session_id
        session = await create_session(session.user_id, request)
        await delete_session(old_session_id)

    request.state.session_id = session.session_id
    request.state.created_at = session.created_at
    request.state.expires_at = session.expires_at
    request.state.user_id = session.user_id
    request.state.permissions = session.permissions

    return await call_next(request)